### DISCUSSION TITLE
Github 지표 수집 자동화 시트

### DISCUSSION THREAD

**Discussion by zion830** [euphony-io/euphony#199] [General]
Github API랑 스프레드시트를 연동해서 만들어봤습니다.
* 수집하는 데이터 : PR, PR 코멘트(총 합계만), 머지된 커밋, 이슈
* organization-level discussion은 github api에서 지원해주지 않아서 직접 수집해야 할듯 합니다 ㅠ
https://docs.google.com/spreadsheets/d/10GsukEC4I4jnNHx-CWXS-IO_optkF_02UVV5L5oekY8/edit?usp=sharing

## 이용 방법
1. euphony-io 하위에 있는 repository 이름 입력 (ex. euphony, order-with, waple 등)
2. 수집 시작일~종료일 선택 (날짜는 무조건 `YYYY.MM.DD` 형식이어야 합니다)
3. 시트가 수정될때마다 자동으로 갱신 스크립트가 돌아갑니다. 속도가 좀 느려서 5초 정도 기다려야 데이터가 갱신됩니다.
    - 수동으로 갱신도 가능합니다. [**확장 프로그램 - 매크로 - ⭐️ 지표 수집 시작**]을 클릭하면 됩니다.
<img width="549" alt="스크린샷 2022-08-22 오후 11 05 53" src="https://user-images.githubusercontent.com/35446215/185940856-3fa65468-c949-48f6-abbe-faa9cefab22f.png">

## 주의사항
* 시간당 API Limit가 있어서 갱신을 과도하게 하면 데이터가 안받아와집니다.
* 갱신후 표가 비었거나 상태 알리미에 에러 메세지가 뜨면 수동으로 갱신해주세요!
* 노란색 칸 이외의 영역을 수정하면 시트가 고장날 수 있습니다

## 구현 안된 부분
* 개인별 PR 코멘트 갯수는 구현이 아직 안되었습니다
* commit api에서 첫번째 author 정보만 내려와서 개인별 commit count에 co-author까지 반영하려면 커밋 메세지 파싱 작업이 필요합니다

**Comment by designe** [comment]
@zion830 우와 👍 대단합니다 정말루 짱이네요!!!

**Comment by zion830** [comment]
집계 기준에 대해 추가로 덧붙입니다~!

## 1. 시작 기간~종료 기간은 created_at을 기준으로 필터링 합니다.
8월 1일\~2일로 기준을 걸었다면, 말 그대로 8월 1일\~2일 사이 새로 만들어진 것들만 집계됩니다.
7월 29일에 생성된 PR이 8월 1일에 업데이트 되었더라도 집계엔 포함되지 않습니다.

## 2. PR comments 집계 기준
가장 오른쪽에 있는 `PR comments`는 **PR list를 내려주는 api 응답에 들어있는 `comments`라는 값의 합계**입니다.
처음엔 당연히 이 값이 PR마다 달려있는 모든 comment 갯수인줄 알고 구현했는데요.
실제로는 깃허브 PR 탭에서 보이는 코멘트 갯수보다 숫자가 적게 나옵니다.
(response 응답값이 각각 뭐하는 값인지 공식문서에서 하나하나 설명해주지 않아서 comments가 대체 뭔지 직접 세보고 알게 됐습니다ㅎㅎ)

예를들어 최근에 제가 open한 #194 는 깃허브에서 보면 코멘트가 3개인 것으로 표기되지만, 시트에서는 도혁님이 남기신 코멘트 1개만 집계됩니다. 
**`xxx commented` 이런식으로 쓰여있는 일반 코멘트만 집계가 되기 때문입니다.**
코드 리뷰나 approve 하면서 남기는 메세지는 `xxx left a comment` 이렇게 표기되는데, 이것들은 포함되지 않습니다.

<img width="469" alt="스크린샷 2022-08-23 오전 1 00 01" src="https://user-images.githubusercontent.com/35446215/185966266-24c055be-7a61-4105-ad72-ebbbe284fa2b.png">

코드 리뷰 갯수까지 완벽하게 포함하려면 PR id를 가지고 review list를 받아오는 API를 한번 더 이용해야 합니다.
만약에 검색 기간이 길어서 조회해야하는 PR이 50개면 API를 50번 쏴야해서 속도도 그렇고 api limit을 금방 돌파하게 됩니다.
개별 comment 갯수를 생략한 것도 동일한 이유입니다. 어떤 사람이 코멘트를 작성했는지 확인하려면 PR 별로 comment list를 받아와야 하기 때문입니다

## 3줄 요약
1. 처음 만들어진 날짜를 기준으로 필터링 됩니다.
2. PR comments엔 코드리뷰나 approve 메세지는 집계되지 않습니다. `xxx commented`로 표기되는 일반 코멘트만 집계됩니다.
3. 따라서 PR comments 데이터에는 크게 의미를 두지 않는게 좋을거 같습니다.
